I can not receive the values of "byte[64][]" returned by the contract.

It is the continuation of #47.
Thank you for the correspondence of #47.

When I tried it with the latest source, I can compile it, but I get an InvalidOutputType error.
I tried several while changing the type, but it did not work.

contract interface:
```
function getValues() public constant returns (uint[], byte[64][], address[])
```

rust source:
```
let result = contract.query("getValues", (), None, Options::default());
let a = result.wait();
match a {
    Ok(_) => {
        let (v1, v2, v3): (Vec<U256>, Vec<Vec<u8>>, Vec<Address>) = a.unwrap();
        println!("{:?} {:?} {:?}", v1, v2, v3);
    }
    Err(e) => println!("{:?}", e),
};
```

Try 1:
```
let (v1, v2, v3): (Vec<U256>, Vec<Vec<u8>>, Vec<Address>) = a.unwrap();
```
```
InvalidOutputType("Expected `bytes`, got FixedArray([FixedBytes([206]), FixedBytes([72]), FixedBytes([241]), FixedBytes([97]), FixedBytes([207]), FixedBytes([128]), FixedBytes([204]), FixedBytes([70]), FixedBytes([118]), FixedBytes([206]), FixedBytes([78]), FixedBytes([231]), FixedBytes([240]), FixedBytes([253]), FixedBytes([6]), FixedBytes([167]), FixedBytes([83]), FixedBytes([97]), FixedBytes([140]), FixedBytes([94]), FixedBytes([69]), FixedBytes([6]), FixedBytes([147]), FixedBytes([168]), FixedBytes([87]), FixedBytes([143]), FixedBytes([48]), FixedBytes([134]), FixedBytes([203]), FixedBytes([234]), FixedBytes([137]), FixedBytes([164]), FixedBytes([184]), FixedBytes([9]), FixedBytes([192]), FixedBytes([144]), FixedBytes([96]), FixedBytes([23]), FixedBytes([52]), FixedBytes([205]), FixedBytes([78]), FixedBytes([122]), FixedBytes([134]), FixedBytes([64]), FixedBytes([80]), FixedBytes([76]), FixedBytes([37]), FixedBytes([106]), FixedBytes([206]), FixedBytes([151]), FixedBytes([106]), FixedBytes([107]), FixedBytes([192]), FixedBytes([75]), FixedBytes([183]), FixedBytes([11]), FixedBytes([217]), FixedBytes([134]), FixedBytes([250]), FixedBytes([98]), FixedBytes([112]), FixedBytes([183]), FixedBytes([106]), FixedBytes([225])])")
```

Try 2:
```
let (v1, v2, v3): (Vec<U256>, Vec<Vec<H256>>, Vec<Address>) = a.unwrap();
```
```
InvalidOutputType("Expected `H256`, got [206]")
```

Try 3:
```
let (v1, v2, v3): (Vec<U256>, Vec<Vec<[H256; 1]>>, Vec<Address>) = a.unwrap();
```
```
InvalidOutputType("Expected `FixedArray(1)`, got FixedBytes([206])")
```

How can I receive the values?
Thank you.

Thank you for fixing.
With #51 source, this worked fine!

```
let (v1, v2, v3): (Vec<U256>, Vec<Vec<Vec<u8>>>, Vec<Address>) = a.unwrap();
```

However, this is a compile error.
```
let (v1, v2, v3): (Vec<U256>, Vec<[[u8; 1]; 64]>, Vec<Address>) = a.unwrap();
```
```
error[E0277]: the trait bound `for<'de> [[u8; 1]; 64]: serde::Deserialize<'de>` is not satisfied
   --> src\test.rs:123:24
    |
123 |         let a = result.wait();
    |                        ^^^^ the trait `for<'de> serde::Deserialize<'de>` is not implemented for `[[u8; 1]; 64]`
    |
    = help: the following implementations were found:
              <&'a [u8] as serde::Deserialize<'de>>
              <[T; 0] as serde::Deserialize<'de>>
              <[T; 1] as serde::Deserialize<'de>>
              <[T; 2] as serde::Deserialize<'de>>
            and 30 others
    = note: required because of the requirements on the impl of `for<'de> serde::Deserialize<'de>` for `std::vec::Vec<[[u8; 1]; 64]>`
    = note: required because of the requirements on the impl of `for<'de> serde::Deserialize<'de>` for `(std::vec::Vec<web3::types::U256>, std::vec::Vec<[[u8; 1]; 64]>, std::vec::Vec<web3::types::H160>)`
    = note: required because of the requirements on the impl of `serde::de::DeserializeOwned` for `(std::vec::Vec<web3::types::U256>, std::vec::Vec<[[u8; 1]; 64]>, std::vec::Vec<web3::types::H160>)`
    = note: required because of the requirements on the impl of `jsonrpc_core::<unnamed>::Future` for `web3::contract::result::QueryResult<(std::vec::Vec<web3::types::U256>, std::vec::Vec<[[u8; 1]; 64]>, std::vec::Vec<web3::types::H160>), web3::transports::http::FetchTask<fn(&str) -> std::result::Result<serde_json::Value, web3::Error>>>`
```

Perhaps, I think that it is a problem that there is no "64" here.
https://github.com/serde-rs/serde/blob/master/serde/src/de/impls.rs#L723

Thank you.

--- a/src/contract/tokens.rs
+++ b/src/contract/tokens.rs
@@ -186,6 +186,7 @@ impl Tokenizable for Vec<u8> {
   fn from_token(token: Token) -> Result<Self, Error> {
     match token {
       Token::Bytes(data) => Ok(data),
+      Token::FixedBytes(data) => Ok(data),
       other => Err(Error::InvalidOutputType(format!("Expected `bytes`, got {:?}", other))),
     }
   }
@@ -207,8 +208,31 @@ impl<T: Tokenizable> Tokenizable for Vec<T> {
   }
 }
 
-macro_rules! impl_fixed_array {
+macro_rules! impl_fixed_types {
   ($num: expr) => {
+    impl Tokenizable for [u8; $num] {
+      fn from_token(token: Token) -> Result<Self, Error> {
+        match token {
+          Token::FixedBytes(bytes) => {
+            if bytes.len() != $num {
+              return Err(Error::InvalidOutputType(format!("Expected `FixedBytes({})`, got FixedBytes({})", $num, bytes.len())));
+            }
+
+            let mut arr = [0; $num];
+            for i in 0..$num {
+              arr[i] = bytes[i];
+            }
+            Ok(arr)
+          },
+          other => Err(Error::InvalidOutputType(format!("Expected `FixedBytes({})`, got {:?}", $num, other))),
+        }
+      }
+
+      fn into_token(self) -> Token {
+        Token::FixedBytes(self.to_vec())
+      }
+    }
+
     impl<T: Tokenizable + Clone> Tokenizable for [T; $num] {
       fn from_token(token: Token) -> Result<Self, Error> {
         match token {
@@ -239,19 +263,19 @@ macro_rules! impl_fixed_array {
   }
 }
 
-impl_fixed_array!(1);
-impl_fixed_array!(2);
-impl_fixed_array!(3);
-impl_fixed_array!(4);
-impl_fixed_array!(5);
-impl_fixed_array!(8);
-impl_fixed_array!(16);
-impl_fixed_array!(32);
-impl_fixed_array!(64);
-impl_fixed_array!(128);
-impl_fixed_array!(256);
-impl_fixed_array!(512);
-impl_fixed_array!(1024);
+impl_fixed_types!(1);
+impl_fixed_types!(2);
+impl_fixed_types!(3);
+impl_fixed_types!(4);
+impl_fixed_types!(5);
+impl_fixed_types!(8);
+impl_fixed_types!(16);
+impl_fixed_types!(32);
+impl_fixed_types!(64);
+impl_fixed_types!(128);
+impl_fixed_types!(256);
+impl_fixed_types!(512);
+impl_fixed_types!(1024);
 
 #[cfg(test)]
 mod tests {
@@ -276,7 +300,30 @@ mod tests {
     let _pair: (U256, bool) = output();
     let _vec: Vec<U256> = output();
     let _array: [U256; 4] = output();
+    let _bytes: Vec<[[u8; 1]; 64]> = output();
 
     let _mixed: (Vec<Vec<u8>>, [U256; 4], Vec<U256>, U256) = output();
+  }
+
+  #[test]
+  fn should_decode_array_of_fixed_bytes() {
+    // byte[8][]
+    let tokens = vec![
+      Token::FixedArray(vec![
+        Token::FixedBytes(vec![1]),
+        Token::FixedBytes(vec![2]),
+        Token::FixedBytes(vec![3]),
+        Token::FixedBytes(vec![4]),
+        Token::FixedBytes(vec![5]),
+        Token::FixedBytes(vec![6]),
+        Token::FixedBytes(vec![7]),
+        Token::FixedBytes(vec![8]),
+      ])
+    ];
+    let data: [[u8; 1]; 8] = Detokenize::from_tokens(tokens).unwrap();
+    assert_eq!(data[0][0], 1);
+    assert_eq!(data[1][0], 2);
+    assert_eq!(data[2][0], 3);
+    assert_eq!(data[7][0], 8);
   }
 }

--- a/src/contract/tokens.rs
+++ b/src/contract/tokens.rs
@@ -219,9 +219,7 @@ macro_rules! impl_fixed_types {
             }
 
             let mut arr = [0; $num];
-            for i in 0..$num {
-              arr[i] = bytes[i];
-            }
+            arr.copy_from_slice(&bytes);
             Ok(arr)
           },
           other => Err(Error::InvalidOutputType(format!("Expected `FixedBytes({})`, got {:?}", $num, other))),

